cyipopt/utils.py

```diff
@@ -91,3 +91,33 @@ def generate_deprecation_warning_msg(what,
            f"deprecated in CyIpopt. Please replace all uses and use "
            f"'{new_name}' going forward.")
     return msg
+
+
+class FunctionWithApproxJacobian(object):
+    def __init__(self, func, epsilon):
+        self._func = func
+        self.epsilon = epsilon
+        self.value_cache = {}
+
+    def __call__(self, x, *args, **kwargs):
+        key = tuple(x)
+        if not key in self.value_cache:
+            value = self._func(x, *args, **kwargs)
+            if np.any(np.isnan(value)):
+                print("Warning! nan function value encountered at {0}".format(x))
+            self.value_cache[key] = value
+        return self.value_cache[key]
+
+    def func(self, x, *args, **kwargs):
+        return self(x, *args, **kwargs)
+
+    def jac(self, x, *args, **kwargs):
+        x0 = np.asfarray(x)
+        dxs = np.zeros((len(x0), len(x0) + 1))
+        for i in range(len(x0)):
+            dxs[i, i + 1] = self.epsilon
+        results = [self(*(x0 + dxs[:, i], ) + args, **kwargs) for i in range(len(x0) + 1)]
+        jac = np.zeros([len(x0), len(np.atleast_1d(results[0]))])
+        for i in range(len(x0)):
+            jac[i] = (results[i + 1] - results[0]) / self.epsilon
```

Comment:
Using a forward finite difference with a constant epsilon is liable to produce a large error. If we are going to support approximating numerical Jacobians on the user's behalf I think we need to make a good effort to ensure that this is accurate. I'd advocate using central differencing and dynamically computing the optimal epsilon to minimise the expected error (e.g. based on the method from [Gill, Murray & Wright](https://books.google.co.uk/books/about/Practical_Optimization.html?id=G-vEDwAAQBAJ&source=kp_book_description&redir_esc=y)).